### PR TITLE
planner: add cascade basic interface definition and basic task type.

### PR DESCRIPTION
**Pull request by AilinKid**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/51664

Problem Summary:

### What changed and how does it work?
whiteboard_exported_image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  > 
when the scheduler part is on, it will be more easier to test the memo workflow, currently single task couldn't run itself

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/cascades/base/cascadesctx/cascades_ctx.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+// limitations under the License.
+
+package cascadesctx
+
+import (
+	"github.com/pingcap/tidb/pkg/planner/cascades/base"
+	"github.com/pingcap/tidb/pkg/planner/cascades/memo"
+)
+
+// CascadesContext define the yams context as interface, since it will be defined
```
Comment (winoros) [inline pkg/planner/cascades/base/cascadesctx/cascades_ctx.go:22]: Do we still need the word `yams`?

File: pkg/planner/cascades/task/task_opt_group.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"github.com/pingcap/tidb/pkg/planner/cascades/base"
+	"github.com/pingcap/tidb/pkg/planner/cascades/base/cascadesctx"
+	"github.com/pingcap/tidb/pkg/planner/cascades/memo"
+	"github.com/pingcap/tidb/pkg/planner/cascades/util"
+)
+
+var _ base.Task = &OptGroupTask{}
+
+type OptGroupTask struct {
```
Comment (winoros) [inline pkg/planner/cascades/task/task_opt_group.go:26]: Do we need to comment the usage of this struct?

File: pkg/planner/cascades/task/task_opt_group_expression.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	"github.com/pingcap/tidb/pkg/planner/cascades/base"
+	"github.com/pingcap/tidb/pkg/planner/cascades/base/cascadesctx"
+	"github.com/pingcap/tidb/pkg/planner/cascades/memo"
+	"github.com/pingcap/tidb/pkg/planner/cascades/rule"
+	"github.com/pingcap/tidb/pkg/planner/cascades/util"
+)
+
+var _ base.Task = &OptGroupTask{}
+
+type OptGroupExpressionTask struct {
```
Comment (winoros) [inline pkg/planner/cascades/task/task_opt_group_expression.go:27]: ditto

**Review by winoros — APPROVED** [review]
The general changes LGTM
But we may need to take care of the possible memory leak in our later tests.

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**